testnet/src/node.rs

```diff
@@ -338,6 +492,114 @@ impl Node {
+                          burn_block: BurnchainTip,
+                          keychain: &mut Keychain,
+                          mem_pool: &mut MemPoolDB,
+                          burn_fee_cap: u64,
+                          bitcoin_controller: &mut BitcoinRegtestController) -> Option<(BurnchainHeaderHash, StacksBlock)> {
+
+        let stacks_tip = match chain_state.get_stacks_chain_tip().unwrap() {
+            Some(x) => x,
+            None => {
+                warn!("Could not mine new tenure, since no chain tip known.");
+                return None
+            }
+        };
+
+        let stacks_tip_header = match StacksChainState::get_anchored_block_header_info(
+            &chain_state.headers_db, &stacks_tip.burn_header_hash, &stacks_tip.anchored_block_hash).unwrap() {
+            Some(x) => x,
+            None => {
+                warn!("Could not mine new tenure, since could not find header for known chain tip.");
+                return None
+            }
+        };
+
+
+        debug!("Mining tenure on burn_block: {}, stacks tip burn_header_hash: {}",
+              &burn_block.block_snapshot.burn_header_hash,
+              &stacks_tip.burn_header_hash);
+
+        // Generates a proof out of the sortition hash provided in the params.
+        let vrf_proof = keychain.generate_proof(
+            &registered_key.vrf_public_key, 
+            burn_block.block_snapshot.sortition_hash.as_bytes()).unwrap();
+
+        debug!("Generated VRF Proof: {} over {} with key {}",
+              vrf_proof.to_hex(),
+              &burn_block.block_snapshot.sortition_hash,
+              &registered_key.vrf_public_key.to_hex());
+
+        // Generates a new secret key for signing the trail of microblocks
+        // of the upcoming tenure.
+        let microblock_secret_key = keychain.rotate_microblock_keypair();
+
+        let burn_header_to_poll = &stacks_tip.burn_header_hash;
+        let block_hash_to_poll = &stacks_tip.anchored_block_hash;
+
+        let ratio = StacksWorkScore {
+            burn: burn_block.block_snapshot.total_burn,
+            work: 1 + stacks_tip_header.anchored_header.total_work.work,
+        };
+
+        // relayer is always building a non-genesis block.
+        let mut block_builder = StacksBlockBuilder::from_parent(
+            1, 
+            &stacks_tip_header,
+            &ratio,
+            &vrf_proof, 
+            &microblock_secret_key);
+
+        let mut clarity_tx = block_builder.epoch_begin(chain_state).unwrap();
+        
+
+        // make a coinbase
+
+        let principal = keychain.origin_address().unwrap().into();
+        let nonce = StacksChainState::get_account(&mut clarity_tx, &principal).nonce;
+
+        let coinbase_tx = Node::inner_generate_coinbase_tx(keychain, nonce);
+        block_builder.try_mine_tx(&mut clarity_tx, &coinbase_tx);
+
+        let txs = mem_pool.poll(burn_header_to_poll, &block_hash_to_poll);
```

Comment:
In the PR I'm working on for mining, I don't use `.poll()` because it could stream back an unbound number of headers at once.  But it's probably fine to leave here for now.